src/py42/services/_connection.py

```diff
@@ -0,0 +1,258 @@
+    def __init__(self, connection, prefix):
+        self._connection = connection
+        self._prefix = prefix
+
+    def get_host_address(self):
+        sts_url = self._get_sts_base_url()
+        return sts_url.replace(u"sts", self._prefix, 1)
+
+    def _get_sts_base_url(self):
+        uri = u"/api/ServerEnv"
+        response = self._connection.get(uri)
+
+        response_json = json_lib.loads(response.text)
+        sts_base_url = response_json.get(u"stsBaseUrl")
+
+        if not sts_base_url:
+            raise Py42FeatureUnavailableError(response)
+
+        return sts_base_url
+
+
+class MicroserviceKeyHostResolver(HostResolver):
+    def __init__(self, kv_service, key):
+        self._kv_service = kv_service
+        self._key = key
+
+    def get_host_address(self):
+        return self._kv_service.get_stored_value(self._key).text
+
+
+class Connection(object):
+    def __init__(self, host_resolver, auth=None, session=None):
+        self._host_resolver = host_resolver
+        self._session = session or ROOT_SESSION
+        self._headers = self._session.headers.copy()
+        self._auth = auth
+        self._resolve_lock = Lock()
+        self._host_address = None
+
+    @classmethod
+    def from_host_address(cls, host_address, auth=None, session=None):
+        host_resolver = KnownUrlHostResolver(host_address)
+        return cls(host_resolver, auth=auth, session=session)
+
+    @classmethod
+    def from_microservice_key(cls, kv_service, key, auth=None, session=None):
+        host_resolver = MicroserviceKeyHostResolver(kv_service, key)
+        return cls(host_resolver, auth=auth, session=session)
+
+    @classmethod
+    def from_microservice_prefix(cls, connection, prefix, auth=None, session=None):
+        host_resolver = MicroservicePrefixHostResolver(connection, prefix)
+        return cls(host_resolver, auth=auth, session=session)
+
+    @property
+    def host_address(self):
+        return self._get_host_address()
+
+    def clone(self, host_address):
+        host_resolver = KnownUrlHostResolver(host_address)
+        return Connection(host_resolver, auth=self._auth)
+
+    def get(self, url, **kwargs):
+        return self.request(u"GET", url, **kwargs)
+
+    def options(self, url, **kwargs):
+        return self.request(u"OPTIONS", url, **kwargs)
+
+    def head(self, url, **kwargs):
+        return self.request(u"HEAD", url, **kwargs)
+
+    def post(self, url, data=None, **kwargs):
+        return self.request(u"POST", url, data=data, **kwargs)
+
+    def put(self, url, data=None, **kwargs):
+        return self.request(u"PUT", url, data=data, **kwargs)
+
+    def patch(self, url, data=None, **kwargs):
+        return self.request(u"PATCH", url, data=data, **kwargs)
+
+    def delete(self, url, **kwargs):
+        return self.request(u"DELETE", url, **kwargs)
+
+    def request(
+        self,
+        method,
+        url,
+        params=None,
+        data=None,
+        json=None,
```

Comment:
Python dictionaries are technically not JSON.  JSON is a string.

Reply:
This mirrors the `requests` implementation of `json`.

Reply:
Or is this just a requests thing?